### PR TITLE
feat: [PL-27008]: Added deletion of settings at removed scopes

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
Updated the Setting Creation Job to delete setting records at the scopes which have been removed from the settings.yml file . Added UTs for SettingCreationJob.java
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/35818)

### CODE REVIEW

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r2, 3 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35818)* status: 10 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/ng-settings/BUILD.bazel` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/35818#-N8cGPmM0Q3TS7ODvGPZ:-N8cGPmM0Q3TS7ODvGP_:bghxr81) ([raw file](https://github.com/harness/harness-core/blob/6f63e27f2710a7cb548cc0ae445d525546ecef82/120-ng-manager/modules/ng-settings/BUILD.bazel#L38)):*
> ```Starlark
>         ":module",
>         "//:lombok",
>         "//120-ng-manager:tests",
> ```

Why do we need this?
___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java` line 166 at r3](https://reviewable.io/reviews/harness/harness-core/35818#-N8cHl_rEMcfDc8oh0Dw:-N8cHl_rEMcfDc8oh0Dx:b-5oai5d) ([raw file](https://github.com/harness/harness-core/blob/6f63e27f2710a7cb548cc0ae445d525546ecef82/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java#L166)):*
> ```Java
>     switch (scopeLevel) {
>       case ACCOUNT:
>         settingsService.deleteByOrgIdentifierNullAndProjectIdentifierNullAndIdentifier(identifier);
> ```

can all these be replaced by a single method. eg. deleteByScope(identifier, scopeLevel) which can have logic to create required criteria.

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35818)* status: 14 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/modules/ng-settings/BUILD.bazel` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/35818#-N8cGPmM0Q3TS7ODvGPZ:-N9G_f0OBBkpy4dmbVRu:b-hfpv0l) ([raw file](https://github.com/harness/harness-core/blob/6f63e27f2710a7cb548cc0ae445d525546ecef82/120-ng-manager/modules/ng-settings/BUILD.bazel#L38)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Why do we need this?
</blockquote></details>

removed
___
*[`120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java` line 166 at r3](https://reviewable.io/reviews/harness/harness-core/35818#-N8cHl_rEMcfDc8oh0Dw:-N9G_hQoBPzpSKq3UKK0:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/6f63e27f2710a7cb548cc0ae445d525546ecef82/120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java#L166)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

can all these be replaced by a single method. eg. deleteByScope(identifier, scopeLevel) which can have logic to create required criteria.
</blockquote></details>

Done.

**Comment by tiwaritejas** [comment]
trigger ti0

**Comment by tiwaritejas** [comment]
trigger ti3

**Comment by tiwaritejas** [comment]
trigger ti0

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 4 of 8 files at r3, 8 of 10 files at r4, 2 of 2 files at r5, 5 of 6 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35818)* status: 12 file(s) to review. 1 of 2 approvals obtained.

File: 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/SettingsServiceImpl.java [hunk 1]
```diff
@@ -213,6 +213,26 @@ public SettingConfiguration upsertSettingConfiguration(SettingConfiguration sett
     return settingConfigurationRepository.save(settingConfiguration);
   }
 
+  @Override
+  public void deleteByScopeLevel(ScopeLevel scopeLevel, String identifier) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/SettingsServiceImpl.java:217]: Why are we not considering scope identifiers as well for deletion? Will this not delete all Settings with identifier from all Projects( if its Project level scope). Or that is the requirement here?

File: 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java [hunk 2]
```diff
@@ -64,14 +65,16 @@ public SettingsCreationJob(
...
+      settingsConfig.getSettings().forEach(settingConfiguration -> {
+        if (settingConfiguration.getAllowOverrides() == null) {
+          settingConfiguration.setAllowOverrides(true);
+        }
+      });
+    }
   }
 
-  private void validateConfig(SettingsConfig settingsConfig) {
+  public void validateConfig(SettingsConfig settingsConfig) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java:77]: Are we using this outside class, hence made this public?

**Review by meenakshi-harness — CHANGES_REQUESTED** [review]

File: 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/SettingsServiceImpl.java (hunk 1 above)
Comment (tiwaritejas) [reply 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/services/impl/SettingsServiceImpl.java:217]: Yes, this is expected. This deletion is for the case when the available scope for a setting is changed. Say, PROJECT scope has been removed from the setting definition by setting owners. In this case, we want all setting records at project scope to be deleted.

File: 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java (hunk 2 above)
Comment (tiwaritejas) [reply 120-ng-manager/modules/ng-settings/src/main/java/io/harness/ngsettings/settings/SettingsCreationJob.java:77]: This was done earlier to test the validate method individually in a UT. But, then I went ahead and tested it by invoking the constructor as validations are done during class instantiation. We can make this private again, but I did not find any issue in keeping it public.

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 3 of 8 files at r3, 6 of 10 files at r4, 5 of 6 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35818)* status: :shipit: complete! ✓

**Pull request merged**